commands: optionally filter `listcoins` by status and outpoint

jp1ac4 opened the pull request:
This is to resolve #676.

I've renamed `CoinType` to `CoinStatus` and updated its values.

The `listcoins` command and related functions can be optionally filtered by coin status and/or outpoint.

Opening this as draft to check that I'm on the right track.

Remaining tasks:
- [x] Update tests
- [x] Update API doc
- [x] Possibly use the updated functions elsewhere, e.g. `list_spending_coins()`

darosior commented on src/commands/mod.rs:
```diff
@@ -747,12 +751,15 @@ impl DaemonControl {
-                .map(|b| current_height + 1 >= b.height + height_delta)
-                .unwrap_or(false)
-        });
+        let sweepable_coins = db_conn
+            .coins(Some(&[CoinStatus::Confirmed]), None)
```
That's a behaviour change. Previously we'd query all unspent coins, including unconfirmed ones. I think just including `CoinStatus::Confirmed` too should fix this.

jp1ac4 replied:
Ah yes, I'll include both confirmed and unconfirmed coins here.

jp1ac4 replied:
I've now updated this.

darosior commented on src/database/mod.rs:
```diff
@@ -85,7 +85,11 @@ pub trait DatabaseConnection {
-    fn coins(&mut self, coin_type: CoinType) -> HashMap<bitcoin::OutPoint, Coin>;
+    fn coins(
+        &mut self,
+        statuses: Option<&[CoinStatus]>,
+        outpoints: Option<&[bitcoin::OutPoint]>,
```
What is `Some([])` supposed to mean? If it's not necessary, then let's drop the `Option` and use the empty list as describing the absence of filters?

jp1ac4 replied:
Oh yes I agree, we can remove the `Option` as the empty list achieves the same thing. I'll change that.

For the `listcoins` CLI args, should we still allow the user to omit the args completely or would we require them to pass an empty list?

If we allow to omit args, I don't think it's possible currently to pass only the second optional arg without the first, e.g.
```
listcoins --outpoints "[\"abcd:1\"]"
```
We would need to do
```
listcoins "[]" "[\"abcd:1\"]"
```

darosior commented on src/database/mod.rs:
```diff
@@ -349,12 +357,25 @@ impl Coin {
-    All,
-    Unspent,
+pub enum CoinStatus {
+    Unconfirmed,
+    Confirmed,
+    Spending,
```
Let's have a doc comment for each variant about how they are mutually exclusive?

jp1ac4 replied:
I've now added a comment for each variant.

darosior commented on src/database/sqlite/mod.rs:
```diff
@@ -357,18 +357,75 @@ impl SqliteConn {
+            if !outpoints.is_empty() {
+                let mut cond = "(txid, vout) IN (VALUES ".to_string();
+                for (i, outpoint) in outpoints.iter().enumerate() {
+                    // NOTE: SQLite doesn't know Satoshi decided txids would be displayed as little-endian
+                    // hex.
```
:)

darosior reviewed (commented):
Approach ACK, i think you are on the right track. A couple early comments below from a quick skim. You probably already intend to do that but just to be sure: please add a unit test exercising the various combinations for the `coins` method of the SQLite wrapper.

> Possibly use the updated functions elsewhere, e.g. list_spending_coins()

Oh, yes, that would be neat!

jp1ac4 commented:
I've now made the following changes:
- added unit test for `coins()` method of SQLite wrapper
- updated functional test for `listcoins`
- removed `Option` from parameters (so empty list means no filter)
- consider both confirmed and unconfirmed coins for sweeping
- use `coins()` method in `list_spending_coins()`
- updated docstring of `spend_coins` to refer to "spending" instead of "spent"
- updated `listcoins` entry in API doc

The only doubt I have is whether it should be possible to omit the command-line args or if an empty list should be passed explicitly.

darosior commented:
Any reason for this to still be marked as WIP / draft?

jp1ac4 commented:
> Any reason for this to still be marked as WIP / draft?

Hmm, not really :smile: I have one question about optional command-line args, but otherwise it's ready for review. Will change that now.

darosior reviewed (commented):
Ok from skimming through the diff here it looks good. Excellent test coverage. I'll play with it a bit locally now and investigate wtf is wrong again with our CI.

darosior commented on src/database/sqlite/mod.rs:
```diff
@@ -889,6 +912,315 @@ CREATE TABLE spend_transactions (
+            .map(|c| c.len())
+            .all(|length| length == 1));
+
+            fs::remove_dir_all(tmp_dir).unwrap();
+        }
```
This ought to be outside the scope of the connection object.

```suggestion
        }
    fs::remove_dir_all(tmp_dir).unwrap();
```

That'll fix the Windows CI.

jp1ac4 replied:
Thanks for this! The failing Windows test is now passing :relieved:

darosior commented:
> The only doubt I have is whether it should be possible to omit the command-line args or if an empty list should be passed explicitly.

It should be possible to omit them, i think. And passing no argument or an empty list should have the same behaviour.

darosior commented on src/database/sqlite/mod.rs:
```diff
@@ -889,6 +912,315 @@ CREATE TABLE spend_transactions (
+                && coin[0].outpoint == coin_b.outpoint
+                && coin[1].outpoint == coin_c.outpoint));
+
+            // Now spend second coin, even though it is still unconfirmed.
```
Nice.

darosior commented on src/jsonrpc/api.rs:
```diff
@@ -87,6 +87,41 @@ fn broadcast_spend(control: &DaemonControl, params: Params) -> Result<serde_json
 }
 
+fn list_coins(control: &DaemonControl, params: Option<Params>) -> Result<serde_json::Value, Error> {
+    let statuses_arg = params
+        .clone()
+        .and_then(|p| p.get(0, "statuses").cloned())
+        .and_then(|statuses| statuses.as_array().cloned());
```
That's a lot of `clone()`s. You don't need to allocate here since you just go through it.

```diff
diff --git a/src/jsonrpc/api.rs b/src/jsonrpc/api.rs
index 1432428..f2b2f4b 100644
--- a/src/jsonrpc/api.rs
+++ b/src/jsonrpc/api.rs
@@ -89,9 +89,9 @@ fn broadcast_spend(control: &DaemonControl, params: Params) -> Result<serde_json
 
 fn list_coins(control: &DaemonControl, params: Option<Params>) -> Result<serde_json::Value, Error> {
     let statuses_arg = params
-        .clone()
-        .and_then(|p| p.get(0, "statuses").cloned())
-        .and_then(|statuses| statuses.as_array().cloned());
+        .as_ref()
+        .and_then(|p| p.get(0, "statuses"))
+        .and_then(|statuses| statuses.as_array());
     let mut statuses: Vec<CoinStatus> = vec![];
     if let Some(statuses_arg) = statuses_arg {
         for status_arg in statuses_arg.iter() {
@@ -104,8 +104,9 @@ fn list_coins(control: &DaemonControl, params: Option<Params>) -> Result<serde_j
         }
     }
     let outpoints = params
-        .and_then(|p| p.get(1, "outpoints").cloned())
-        .and_then(|op| op.as_array().cloned())
+        .as_ref()
+        .and_then(|p| p.get(1, "outpoints"))
+        .and_then(|op| op.as_array())
         .and_then(|arr| {
             arr.iter()
                 .map(|entry| {
```

jp1ac4 replied:
Thanks for this :pray: Have updated it now.

darosior commented on src/jsonrpc/api.rs:
```diff
@@ -87,6 +87,41 @@ fn broadcast_spend(control: &DaemonControl, params: Params) -> Result<serde_json
+                .map(|entry| {
+                    entry
+                        .as_str()
+                        .and_then(|e| bitcoin::OutPoint::from_str(e).ok())
+                })
+                .collect::<Option<Vec<bitcoin::OutPoint>>>()
```
We shouldn't be silently ignoring these errors.

jp1ac4 replied:
Ah yes, I've updated this now.

darosior commented on tests/test_rpc.py:
```diff
@@ -73,6 +101,190 @@ def test_listcoins(lianad, bitcoind):
+            )["coins"],
+            key=lambda c: c["outpoint"],
+        )
+    )
```
Great functional test too. A possible good addition is to sanity check the errors returned in case of invalid inputs. For instance make sure we catch an invalid status or an invalid outpoint. This would have caught the fact we actually ignore the error on malformed outpoints as i pointed above.

Example of such a sanity check:
https://github.com/wizardsardine/liana/blob/38d0fafac2b3cbdba57070173dc0136e20fd7019/tests/test_rpc.py#L610-L612

jp1ac4 replied:
That's a good idea thanks. I've added some error checks to the test.

darosior reviewed (commented):
Looks great. Again, very nice test coverage. The only thing i really care about here is to not silently ignore malformed outpoints in the JSONRPC request.

jp1ac4 commented on src/jsonrpc/api.rs:
```diff
@@ -87,6 +87,58 @@ fn broadcast_spend(control: &DaemonControl, params: Params) -> Result<serde_json
+        .as_ref()
+        .and_then(|p| p.get(1, "outpoints"))
+        .and_then(|op| op.as_array());
+    let mut outpoints: Vec<bitcoin::OutPoint> = vec![];
+    if let Some(outpoints_arg) = outpoints_arg {
+        for op_arg in outpoints_arg.iter() {
+            outpoints.push(
+                op_arg
+                    .as_str()
+                    .ok_or_else(|| {
+                        Error::invalid_params(format!(
+                            "Invalid value {} in 'outpoints' parameter.",
+                            op_arg
+                        ))
+                    })
+                    .map(|op| {
+                        bitcoin::OutPoint::from_str(op).map_err(|_| {
+                            Error::invalid_params(format!(
+                                "Invalid value {} in 'outpoints' parameter.",
+                                op_arg
+                            ))
+                        })
+                    })??,
+            );
+        }
+    }
```
There might be a better way of writing this, but this was the best I could get working :blush:

jp1ac4 replied:
I think I've found a better way :nerd_face: Will test it and then commit shortly.

jp1ac4 replied:
I've now updated this.

jp1ac4 commented:
Thanks for your feedback. I've now made the suggested changes.

jp1ac4 commented:
I've just committed an improved way of getting the arguments from command line.

I also fixed a typo in the unit test name: `db_coins_()` -> `db_coins()`.

I don't have any further changes planned.

darosior commented:
ACK 36e04edc8207df2b3b3c9fac78f11700514ee2bd
